### ISSUE TITLE
New Diode2 model's HeatPort does not work

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1823]
**Reported by stefanv on 6 Nov 2015 13:54 UTC**
Unlike the other Electrical.Analog components, the HeatPort in the new Diode2 model doesn't do anything. Instead, the effect of heat is modelled indirectly (and in a non-varying way) via the Vt parameter. Changing the model to activate the HeatPort is trivial (see `// <---` lines in code below):

```mo
model Diode2 "Improved diode model"
    extends Modelica.Electrical.Analog.Interfaces.OnePort;
    extends Modelica.Electrical.Analog.Interfaces.ConditionalHeatPort(T=293.15);
    parameter SIunits.Voltage Vf = 0.7 "Forward voltage";
    parameter SIunits.Current Ids = 1.e-13 "Reverse saturation current";
    parameter SIunits.Resistance Rs = 16 "Ohmic resistance";
    parameter SIunits.Voltage Vt = Modelica.Constants.R * T
				 / Modelica.Constants.F
	"Thermal voltage (kT/q) when useHeatPort=false, 0.026 at normal conditions (around 20 degC)";
    parameter Real N = 1 "Emission coefficient";
    parameter SIunits.Voltage Bv = 100 "Reverse breakdown voltage";
    parameter SIunits.Conductance Gp = 1e-6
	"Parallel conductance for numerical stability";
    SIunits.Voltage vd "Voltage across pure diode part";
    SIunits.Current id "diode current";
    SIunits.Voltage VdMax "Linear continuation threshold";  // <--- These are now variables
    SIunits.Current iVdMax "Current at threshold";          // <---
    SIunits.Conductance diVdMax "Conductance at threshold"; // <---
    SIunits.Voltage VtUse;                                  // <---
equation
    /* Parameters which depend on a possibly variable temperature. */
    VtUse = if useHeatPort // <---
	    then Modelica.Constants.R * T_heatPort / Modelica.Constants.F // <---
	    else Vt; // <---
    VdMax = Vf + (N*VtUse); // <--- All uses of Vt after here changed to VtUse
    iVdMax = Ids * (exp(VdMax/(N*VtUse)) - 1); // <---
    diVdMax = Ids * exp(VdMax/(N*VtUse))/(N*VtUse); // <---

    id = smooth(1,
      if vd < -Bv / 2 then
	  /* Lower half of reverse biased region including breakdown. */
	  -Ids * (exp(-(vd+Bv)/(N*VtUse)) + 1 - 2*exp(-Bv/(2*N*VtUse)))
      elseif vd < VdMax then
	  /* Upper half of reverse biased region, and forward biased region
	     before conduction. */
	  Ids * (exp(vd/(N*VtUse)) - 1)
      else
	  /* Forward biased region after conduction. */
	  iVdMax + (vd - VdMax) * diVdMax);

    v = vd + id * Rs;
    i = id + v * Gp;
    LossPower = i * v;

    assert(Bv>0, "Bv must be greater than zero", AssertionLevel.error);
    assert(Vf>0, "Vf must be greater than zero", AssertionLevel.error);
    assert(Vt>0, "Vt must be greater than zero", AssertionLevel.error);
    annotation (...);
end Diode2;
```

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1823

**Comment by modelica-trac-importer** [comment]
**Comment by majetta on 10 Dec 2015 14:21 UTC**
changes as described in that ticket added to Modelica.Electrical.Analog.Semiconductors.Diode2 in 0f2a789f6e6701eebcf736afdb8574e23f189876

**Comment by modelica-trac-importer** [comment]
**Changelog modified by majetta on 10 Dec 2015 14:21 UTC**
temperature dependency added to Modelica.Electrical.Analog.Semiconductors.Diode2

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 10 Dec 2015 14:27 UTC**
Please remove the `uses` annotation from your commit.

**Comment by modelica-trac-importer** [comment]
**Comment by majetta on 10 Dec 2015 15:09 UTC**
uses-annotation and <span>-tag removed in eb32e1257b5d7ae73036c9c9da40b60e71c29bed

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 11 Dec 2015 06:56 UTC**
A student of mine noticed that also the `Modelica.Electrical.Analog.Semiconductors.ZDiode` model has a heatport but the behaviour of the z-diode is no influenced by the temperature. Are there any plans to introduce this?

**Comment by modelica-trac-importer** [comment]
**Changelog modified by dietmarw on 11 Dec 2015 06:56 UTC**
Temperature dependency added to Modelica.Electrical.Analog.Semiconductors.Diode2

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 11 Dec 2015 07:18 UTC**
 * Moved `Diode2` up below `Diode` in e1c90ae2bc86982d179c1ff30bd6638e9d6bac8b
 * Replaced `*` in the documentation equation by the more accurate `&middot;` symbol in 4faff469235476539755a498281eb7d9a45083a9.
 * Fixed number format of a parameter in 91e150035aafb69fc04c15582f8c893e02db757c

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 11 Dec 2015 16:07 UTC**
